util/conditions/v1beta2/getter.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta2
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// TODO: Move to the API package.
+const (
+	// NoReasonReported identifies a clusterv1.Condition that reports no reason.
+	NoReasonReported = "NoReasonReported"
+)
+
+// Getter interface defines methods that an API object should implement in order to
+// use the conditions package for getting conditions.
+type Getter interface {
+	// GetV1Beta2Conditions returns the list of conditions for a cluster API object.
+	// Note: GetV1Beta2Conditions will be renamed to GetConditions in a later stage of the transition to V1Beta2.
+	GetV1Beta2Conditions() []metav1.Condition
+}
+
+// Get returns a condition from the object implementing the Getter interface.
+//
+// Please note that Get does not support reading conditions from unstructured objects nor from API types not implementing
+// the Getter interface. Eventually, users can implement wrappers on those types implementing this interface and
+// taking care of aligning the condition format if necessary.
+func Get(sourceObj Getter, sourceConditionType string) *metav1.Condition {
+	// if obj is nil, the requested condition does not exist.
+	if sourceObj == nil {
+		return nil
+	}
+
+	// if obj is a zero value of its type (a pointer to nil), the requested condition does not exist.
+	if ptr := reflect.ValueOf(sourceObj); ptr.Kind() == reflect.Pointer {
+		elem := ptr.Elem()
+		if !elem.IsValid() {
+			return nil
+		}
+	}
```

Comment:
Just realized now that this check is very different from what we have as a util:
```go
// IsNil returns an error if the passed interface is equal to nil or if it has an interface value of nil.
func IsNil(i interface{}) bool {
	if i == nil {
		return true
	}
	switch reflect.TypeOf(i).Kind() {
	case reflect.Ptr, reflect.Map, reflect.Chan, reflect.Slice, reflect.Interface, reflect.UnsafePointer, reflect.Func:
		return reflect.ValueOf(i).IsValid() && reflect.ValueOf(i).IsNil()
	}
	return false
}
```